### PR TITLE
Fix Async_Compute with SSR build - 1395932

### PR DESCRIPTION
**Pull request by skhiat**
Fix when we use the -build- a version with SSR + Async Compute Shader:
image

Discussion ref:
https://unity.slack.com/archives/CM6VCD9C7/p1616763855031800

This PR should fix this PR:
image

To the tester:
Could you validate that the issue is not here any more, I wasn't able to produce a proper build.

### CODE REVIEW

**Review by TomasKiniulis — COMMENTED** [review]
SSR errors are resolved, but Dx12 player build does not run yet. Not sure if related directly to SSR or should be solved with another PR.

The following errors remain:
```
Render Graph Execution error
UnityEngine.StackTraceUtility:ExtractStackTrace () (at C:/buildslave/unity/build/Runtime/Export/Scripting/StackTrace.cs:37)
...
```

```
IndexOutOfRangeException: Index was outside the bounds of the array.
  at UnityEngine.Rendering.DynamicArray`1[T].get_Item (System.Int32 index) [0x00026] in C:\GitHub\Graphics2\Graphics\com.unity.render-pipelines.core\Runtime\Common\DynamicArray.cs:258 
  at UnityEngine.Experimental.Rendering.RenderGraphModule.RenderGraph.UpdateResourceAllocationAndSynchronization () [0x00206] in C:\GitHub\Graphics2\Graphics\com.unity.render-pipelines.core\Runtime\RenderGraph\RenderGraph.cs:1148 
  at UnityEngine.Experimental.Rendering.RenderGraphModule.RenderGraph.CompileRenderGraph () [0x00047] in C:\GitHub\Graphics2\Graphics\com.unity.render-pipelines.core\Runtime\RenderGraph\RenderGraph.cs:1248 
...
```

**Comment by kecho** [comment]
Is there any reason why we couldnt do 
1) one pass to just clear the textures
2) the async pass using the cleared  prepared textures?
I say this because there is a performance cost, significant, into clearing textures manually. In console the rule is .3ms for a 1080p target. If half res, .15 or .1 ms. 
In Async we would be introducing bandwidth cost, which is no bueno if we have also bandwidth in flight.

Now if you really want to push for a manual clear, we need to measure this at least in ps4. Lmk if you need help profiling this.

**Comment by tlaedre** [comment]
> Is there any reason why we couldnt do
> 
> 1. one pass to just clear the textures
> 2. the async pass using the cleared  prepared textures?
>    I say this because there is a performance cost, significant, into clearing textures manually. In console the rule is .3ms for a 1080p target. If half res, .15 or .1 ms.
>    In Async we would be introducing bandwidth cost, which is no bueno if we have also bandwidth in flight.
> 
> Now if you really want to push for a manual clear, we need to measure this at least in ps4. Lmk if you need help profiling this.

Wouldn't this require dedicated resources instead of shared rendergraph resources to work @kecho ?

**Comment by kecho** [comment]
> > Is there any reason why we couldnt do
> > 
> > 1. one pass to just clear the textures
> > 2. the async pass using the cleared  prepared textures?
> >    I say this because there is a performance cost, significant, into clearing textures manually. In console the rule is .3ms for a 1080p target. If half res, .15 or .1 ms.
> >    In Async we would be introducing bandwidth cost, which is no bueno if we have also bandwidth in flight.
> > 
> > Now if you really want to push for a manual clear, we need to measure this at least in ps4. Lmk if you need help profiling this.
> 
> Wouldn't this require dedicated resources instead of shared rendergraph resources to work @kecho ?

Hmm not following? it would still import the textures as required.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs [hunk 1]
```diff
@@ -1187,18 +1190,79 @@ void UpdateSSRConstantBuffer(HDCamera hdCamera, ScreenSpaceReflection settings,
...
+                        passData.ssrAccumPrev = builder.WriteTexture(renderGraph.ImportTexture(hdCamera.GetPreviousFrameRT((int)HDCameraFrameHistoryType.ScreenSpaceReflectionAccumulation)));
 
+                        builder.SetRenderFunc(
+                            (RenderSSRPassData data, RenderGraphContext ctx) =>
+                            {
+                                var cs = data.ssrCS;
+                                passData.clearBuffer2DCS = m_ClearBuffer2DCS;
+                                passData.clearBuffer2DKernel = m_ClearBuffer2DKernel;
+
+#if true
```
Comment (kecho) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs:1224]: I would remove this

**Review by kecho — CHANGES_REQUESTED** [review]
Id say clean up those #if true and then it can be approved!

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs (hunk 1 above)
Comment (skhiat) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs:1224]: Fixed, tmp code

**Comment by tlaedre** [comment]
> > > Is there any reason why we couldnt do
> > > 
> > > 1. one pass to just clear the textures
> > > 2. the async pass using the cleared  prepared textures?
> > >    I say this because there is a performance cost, significant, into clearing textures manually. In console the rule is .3ms for a 1080p target. If half res, .15 or .1 ms.
> > >    In Async we would be introducing bandwidth cost, which is no bueno if we have also bandwidth in flight.
> > > 
> > > Now if you really want to push for a manual clear, we need to measure this at least in ps4. Lmk if you need help profiling this.
> > 
> > 
> > Wouldn't this require dedicated resources instead of shared rendergraph resources to work @kecho ?
> 
> Hmm not following? it would still import the textures as required.

But you can't actually _use_ those imported textures outside of the render func, that would defeat the whole purpose of the render graph.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs [hunk 2]
```diff
@@ -1187,18 +1190,66 @@ void UpdateSSRConstantBuffer(HDCamera hdCamera, ScreenSpaceReflection settings,
                 if (colorPyramidRT == null)
                     return renderGraph.defaultResources.blackTextureXR;
 
-                using (var builder = renderGraph.AddRenderPass<RenderSSRPassData>("Render SSR", out var passData))
+                bool usePBRAlgo = !transparent && settings.usedAlgorithm.value == ScreenSpaceReflectionAlgorithm.PBRAccumulation;
+                if (usePBRAlgo)
                 {
-                    builder.EnableAsyncCompute(hdCamera.frameSettings.SSRRunsAsync());
+                    using (var builder = renderGraph.AddRenderPass<RenderSSRPassData>("Clear SSR Buffers", out var passData))
```
Comment (tlaedre) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs:1196]: Shouldn't this pass also run async compute?

**Review by TomasKiniulis — CHANGES_REQUESTED** [review]
>SSR errors are resolved, but Dx12 player build does not run correctly yet. Not sure if related directly to SSR or should be solved with another PR.
 The following errors remain:
Render Graph Execution error
UnityEngine.StackTraceUtility:ExtractStackTrace () (at C:/buildslave/unity/build/Runtime/Export/Scripting/StackTrace.cs:37)
...
IndexOutOfRangeException: Index was outside the bounds of the array.
  at UnityEngine.Rendering.DynamicArray`1[T].get_Item (System.Int32 index) [0x00026] in C:\GitHub\Graphics2\Graphics\com.unity.render-pipelines.core\Runtime\Common\DynamicArray.cs:258 
  at UnityEngine.Experimental.Rendering.RenderGraphModule.RenderGraph.UpdateResourceAllocationAndSynchronization () [0x00206] in C:\GitHub\Graphics2\Graphics\com.unity.render-pipelines.core\Runtime\RenderGraph\RenderGraph.cs:1148 
  at UnityEngine.Experimental.Rendering.RenderGraphModule.RenderGraph.CompileRenderGraph () [0x00047] in C:\GitHub\Graphics2\Graphics\com.unity.render-pipelines.core\Runtime\RenderGraph\RenderGraph.cs:1248 
...

Still hitting the same rendergraph issue.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs [hunk 3]
```diff
@@ -1288,9 +1342,19 @@ void UpdateSSRConstantBuffer(HDCamera hdCamera, ScreenSpaceReflection settings,
...
+                                    ctx.cmd.DispatchCompute(data.clearBuffer2DCS, data.clearBuffer2DKernel, HDUtils.DivRoundUp(data.width, 8), HDUtils.DivRoundUp(data.height, 8), data.viewCount);
+                                }
                                 if (data.previousAccumNeedClear || data.debugDisplaySpeed)
-                                    CoreUtils.SetRenderTarget(ctx.cmd, data.ssrAccumPrev, ClearFlag.Color, Color.clear);
+                                {
+                                    ctx.cmd.SetComputeTextureParam(data.clearBuffer2DCS, data.clearBuffer2DKernel, HDShaderIDs._Buffer2D, data.ssrAccumPrev);
+                                    ctx.cmd.SetComputeVectorParam(data.clearBuffer2DCS, HDShaderIDs._ClearValue, Color.clear);
+                                    ctx.cmd.SetComputeVectorParam(data.clearBuffer2DCS, HDShaderIDs._BufferSize, new Vector4((float)data.width, (float)data.height, 0.0f, 0.0f));
+                                    ctx.cmd.DispatchCompute(data.clearBuffer2DCS, data.clearBuffer2DKernel, HDUtils.DivRoundUp(data.width, 8), HDUtils.DivRoundUp(data.height, 8), data.viewCount);
+                                }
```
[the remark this answers is not in the document]
Comment (skhiat) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.LightLoop.cs:1357]: data.ssrAccum & data.ssrAccumPrev are different texture

**Comment by sebastienlagarde** [comment]
replicate here: https://github.cds.internal.unity3d.com/unity/unity/pull/8592

**Pull request closed without merging**